[wip]: Efficient Token program (#2)
febo/solana-improvement-documents: proposals/XXXX-efficient-token-program.md

---
simd: 'XXXX'

> title: Efficient Token program

Reviewer: nit: May as well include "P-Token" in the title if we're confident with the name :smile:

status: Draft
created: 2025-01-26
feature: (fill in with feature tracking issues once accepted)
supersedes: (optional - fill this in if the SIMD supersedes a previous SIMD)
superseded-by: (optional - fill this in if the SIMD is superseded by a subsequent
 SIMD)
extends: (optional - fill this in if the SIMD extends the design of a previous
 SIMD)
---

## Summary

Replace the current version of SPL Token (`TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA`) program by a CU-optimized one (`p-token`).

## Motivation

About `~10%` of block compute units is used by the Token program instructions. Decreasing the CUs used by Token program instruction creates block space for other transactions to be executed, i.e., less CUs consumed by Token program, more CUs for other transactions.

As an example, if we reduce the CUs consumed by Token instructions to `1/20th` of their current value, `10%` of block CUs utilized becomes `0.5%`, resulting in a `9.5%` block space gain.

Additionally, there are benefits to downstream programs:
* Better composability since using the Token program instructions require less CUs.
* Cheaper (in terms of CU) cross-program invocations.

## New Terminology

N/A.

> ## Detailed Design

Reviewer: Under the detailed design, we need to spell out the work to be done by validator implementations.

Specifically, when the accompanying feature is enabled, the runtime must replace the contents of the account at `Tokenkeg...` with the contents of `$OTHER_ACCOUNT_ADDRESS`. 

But since we can't use BPF Loader v2 anymore, we'll need to swap `Tokenkeg...` with a program using the upgradeable loader. This means that we'll also need to replace the contents of `$TOKENKEG_PROGRAM_DATA_ADDRESS` with the contents of `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS`, where the program data address is defined as `PDA([program_id], upgradeable_loader_id)`.

Code reference at https://github.com/anza-xyz/agave/blob/74b92ed36a70026629a0b303c2eeeff1e8e49f86/sdk/loader-v3-interface/src/lib.rs#L26

Author: The current `Tokenkeg...` does not have a `$TOKENKEG_PROGRAM_DATA_ADDRESS`, right? The program data is in the actual `Tokenkeg...` account since it is a BPF Loader v2 program account. 

If that is the case, then we can leave the `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS` as it is, and replace `Tokenkeg...` contents with `$OTHER_ACCOUNT_ADDRESS` – the new one will already be referencing the `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS`.

Author: Oooh, I see, `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS` won't have the correct derivation.

Below is a sample of the CU efficient gained by `p-token` compared to the current SPL Token program.

| Instruction                | CU (`p-token`) | CU (`spl-token`) |
|----------------------------|----------------|------------------|
| `InitializeMint`           | 100            | 2967             |
| `InitializeAccount`        | 170            | 4527             |
| `InitializeMultisig`       | 190            | 2973             |
| `Transfer`                 | 153            | 4645             |
| `Approve`                  | 124            | 2904             |
| `Revoke`                   | 97             | 2677             |
| `SetAuthority`             | 126            | 3167             |
| `MintTo`                   | 154            | 4538             |
| `Burn`                     | 168            | 4753             |
| `CloseAccount`             | 147            | 2916             |
| `FreezeAccount`            | 136            | 4265             |
| `ThawAccount`              | 136            | 4267             |

## Alternatives Considered

> N/A.

Reviewer: You could specify as an alternative that deploying p-token to a new address and getting people to transition to that program, but people would be very slow to adopt the new program, if they ever adopt it at all.

> ## Impact

Reviewer: nit: can you move the table with the new CU numbers down to this section?

The main impact is freeing up block CUs, allowing more transactions to be packed in a block; dapp developers benefit since interacting with the Token program will consume significantly less CUs.

## Security Considerations

> `p-token` must be guaranteed to follow the same instructions and accounts layout, as well as have the same behaivour than the current Token implmentation.

Reviewer: nit typos
```suggestion
`p-token` must be guaranteed to follow the same instructions and accounts layout, as well as have the same behaviour as the current Token implementation.
```

> Any potential risk will be mitigated by extensive fixture testing, formal verification and audits.

Reviewer: You could also spell out here that due to the potentially huge economic consequences of this change, the feature will be put to a validator vote.

## Drawbacks (Optional)

> N/A.

Reviewer: We could say that as a drawback, it requires breaking consensus on a non-native program. However, this has been done in the past many times for SPL Token to fix bugs and add new features.

## Motivation

About `~10%` of block compute units is used by the Token program instructions. Decreasing the CUs used by Token program instructions creates block space for other transactions to be executed – i.e., less CUs consumed by Token program, more CUs for other transactions.

As an example, if we reduce the CUs consumed by Token instructions to `1/20th` of their current value, `10%` of block CUs utilized becomes `0.5%`, resulting in a `9.5%` block space gain.

Additionally, there are benefits to downstream programs:
* Better composability since using the Token program instructions require less CUs.
* Cheaper (in terms of CU) cross-program invocations.

## New Terminology

N/A.

## Detailed Design

`p-token` is a like-for-like efficient re-implementation of the Token program. It is `no_std` (no heap memory allocations are made in the program) and uses zero-copy access for instruction and account data. Since it follows the same instructions and accounts layout, it does not require any changes to client code – it works as a drop-in replacement.

Apart from the original SPL Token instructions, this proposal adds two additional instructions to the program:

> 1. [`withdraw_excess_lamports`](https://github.com/solana-program/token/blob/main/p-token/src/processor/withdraw_excess_lamports.rs) (instruction discriminator `38`): allow recovering "bricked" SOL from mint (e.g., USDC mint as `~323` SOL in excess) and multisig accounts. The logic of this instruction is the same as the current SPL Token-2022 instruction. 

Reviewer: It's probably best to call out that there could be some economic consequence of unbricking some SOL, and that we haven't calculated how much extra SOL could be freed up in this way.

> 2. [`batch`](https://github.com/solana-program/token/blob/main/p-token/src/processor/batch.rs) (instruction discriminator `255`): enable efficient CPI interaction with the Token program. This is a new instruction that can execute a variable number on Token instructions in a single invocation of the Token program. Therefore, the CPI invoke units (currently `1000` CU) are only consumed once, instead of each CPI instruction – this significantly improves the CUs required to perform multiple Token instructions in a CPI context.

Reviewer: Can you add just a little more about the impact for any DeFi program? Maybe something like: "Almost every DeFi protocol on Solana performs multiple CPIs to the Token program in one instruction. For example, an AMM performs two transfers during swap, or transfers tokens and mints others during an LP deposit. Programs can use the batch instruction for even more CU gains."

Author: Nice! Added the text.

Note that `withdraw_excess_lamports` discriminator matches the value used in SPL Token-2022, while `batch` has a driscriminator value that is not used in either SPL Token nor Token-2022.

The program will be loaded into an account (`ptokNfvuU7terQ2r2452RzVXB3o4GT33yPWo1fUkkZ2`) prior to enabling the feature gate that triggers the replacement.

When the feature gate `ptokSWRqZz5u2xdqMdstkMKpFurauUpVen7TZXgDpkQ` is enabled, the runtime needs to:

* Replace `TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA` program with `ptokNfvuU7terQ2r2452RzVXB3o4GT33yPWo1fUkkZ2` using the Upgradable Loader `v4`.

> Another point for consideration is to whether to keep the current logs of the program or not. Logs are limited to show the name of the instruction being executed, e.g.:

Reviewer: We'll need to take a position for this proposal in order to get pushback / approval. I can't remember what decision we took during previous discussions, but I'm worried it'll be too disruptive to remove the logging.

Let's move this part to "alternatives considered".

```
Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [1]
Program log: Instruction: Transfer
Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 249 of 200000 compute units
Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success
```

> Logging the instruction name consumes around `103` compute units, which in some cases reprensents almost the same amount required to execute the instruction. Note that errors will always be logged, since this is considered good use of CU.

Reviewer: Failed transactions use up all CUs anyway, so we can be as wasteful as we want in failing :wink:

## Alternatives Considered

As an alternative to replace the current version of SPL Token, `p-token` could be deployed to a new address and people can be encouraged to transition to that program. This would hinder its adoption and benefits, since people would be very slow to adopt the new program, if they adopt it at all.

## Impact

The main impact is freeing up block CUs, allowing more transactions to be packed in a block; dapp developers benefit since interacting with the Token program will consume significantly less CUs.

Below is a sample of the CU efficient gained by `p-token` compared to the current SPL Token program.

> | Instruction                | CU (`p-token`) | CU (`p-token`) + logging | CU (`spl-token`) |

Reviewer: Can you add the percent of CUs used for each instruction? For example, for `InitializeMint`, under `p-token`, specify (4%) and under `p-token` + logging specify (7%)? That might make it clearer just how costly the logging is. I'm open to suggestions of how we can present this more clearly

Author: The percentages of the version without logs look very nice. 😊

| `InitializeMultisig`       | 239            | 348                      | 2973             |
| `Transfer`                 | 146            | 249                      | 4645             |
| `Approve`                  | 150            | 268                      | 2904             |
| `Revoke`                   | 124            | 237                      | 2677             |
| `SetAuthority`             | 159            | 261                      | 3167             |
| `MintTo`                   | 144            | 247                      | 4538             |
| `Burn`                     | 202            | 316                      | 4753             |
| `CloseAccount`             | 152            | 255                      | 2916             |
| `FreezeAccount`            | 170            | 287                      | 4265             |
| `ThawAccount`              | 169            | 283                      | 4267             |

## Security Considerations

`p-token` must be guaranteed to follow the same instructions and accounts layout, as well as have the same behaviour than the current Token implementation.

> Any potential risk will be mitigated by extensive fixture testing, formal verification and audits. Since there are potentially huge economic consequences of this change, the feature will be put to a validator vote.

Reviewer: Since the work's been done, let's add a link to the fuzzing and fixture testing already in place, and also specify the current status of the audit / formal verification. The audit is scheduled, and the formal verification is underway, correct?

Author: Updated.

## Detailed Design

`p-token` ([repository](https://github.com/solana-program/token/tree/main/p-token)) is a like-for-like efficient re-implementation of the Token program. It is `no_std` (no heap memory allocations are made in the program) and uses zero-copy access for instruction and account data. Since it follows the same instructions and accounts layout, it does not require any changes to client code – it works as a drop-in replacement.

Apart from the original SPL Token instructions, this proposal adds two additional instructions to the program:

> 1. [`withdraw_excess_lamports`](https://github.com/solana-program/token/blob/main/p-token/src/processor/withdraw_excess_lamports.rs) (instruction discriminator `38`): allow recovering "bricked" SOL from mint (e.g., USDC mint as `~323` SOL in excess) and multisig accounts. The logic of this instruction is the same as the current SPL Token-2022 instruction. There could economic consequences depending on the amount of "unbricked" SOL &mdash; the total amount of SOL that could be freed up this way has not being calculated.

Reviewer: nits typos
```suggestion
1. [`withdraw_excess_lamports`](https://github.com/solana-program/token/blob/main/p-token/src/processor/withdraw_excess_lamports.rs) (instruction discriminator `38`): allow recovering "bricked" SOL from mint (e.g., USDC mint as `~323` SOL in excess) and multisig accounts. The logic of this instruction is the same as the current SPL Token-2022 instruction. There could be economic consequences depending on the amount of "unbricked" SOL &mdash; the total amount of SOL that could be freed up this way has not been calculated.
```

`p-token` is a like-for-like efficient re-implementation of the Token program. It is `no_std` (no heap memory allocations are made in the program) and uses zero-copy access for instruction and account data. Since it follows the same instructions and accounts layout, it does not require any changes to client code – it works as a drop-in replacement.

Apart from the original SPL Token instructions, this proposal adds two additional instructions to the program:

> 1. `withdraw_excess_lamports` (instruction discriminator `38`): allow recovering "bricked" SOL from mint accounts (e.g., USDC mint as `~323` SOL in excess). The logic of this instruction is the same as the current SPL Token-2022 instruction. 

Reviewer: Important note: it also allows people to withdraw extra SOL from SPL-Token multisig accounts

2. `batch` (instruction discriminator `255`): enable efficient CPI interaction with the Token program. This is a new instruction that can execute a variable number on Token instructions in a single invocation of the Token program. Therefore, the CPI invoke units (currently `1000` CU) are only consumed once, instead of each CPI instruction – this significantly improves the CUs required to perform multiple Token instructions in a CPI context.

> Note that `withdraw_excess_lamports` discriminator matches the value used in SPL Token-2022, while `batch` has a driscriminator value that is not used in either SPL Token nor Token-2022.

Reviewer: Confirming that this is true! I went through all of the interfaces that token-2022 implements to be totally sure

> The program and its program data will be loaded into accounts, `$PTOKEN_PROGRAM_ACCOUNT` and `$PTOKEN_PROGRAM_DATA_ACCOUNT` respectively, prior to enablind the feature gate that triggers the replacement.

Reviewer: nit typo
```suggestion
The program and its program data will be loaded into accounts, `$PTOKEN_PROGRAM_ACCOUNT` and `$PTOKEN_PROGRAM_DATA_ACCOUNT` respectively, prior to enabling the feature gate that triggers the replacement.
```

Reviewer: Also, these will need to be specified in the SIMD, similar to the feature gate address.

## Alternatives Considered

As an alternative to replace the current version of SPL Token, `p-token` could be deployed to a new address and people can be encouraged to transition to that program. This would hinder its adoption and benefits, since people would be very slow to adopt the new program, if they adopt it at all.

Another point considered was to whether keep the current logs on the program or not. Logs are limited to show the name of the instruction being executed, e.g.:
```
Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [1]
Program log: Instruction: Transfer
Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 249 of 200000 compute units
Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success
```
Logging the instruction name consumes around `103` compute units, which in some cases reprensents almost the same amount required to execute the instruction, although removing the logs can be too disruptive to anyone relying on them.

## Impact

The main impact is freeing up block CUs, allowing more transactions to be packed in a block; dapp developers benefit since interacting with the Token program will consume significantly less CUs.

> Below is a sample of the CU efficient gained by `p-token` compared to the current SPL Token program. In bracket is the percentage of CU used in relation to the current SPL Token consumption &mdash; the lower the percentage, the better the gains in CU consumption.

Reviewer: I like the addition, it makes the table even clearer
